controllers/bpfman-agent/cl-tc-program.go

```diff
@@ -0,0 +1,349 @@
+	// attach points.
+	attachPointsToRemove := make(map[int]bool)
+
+	var lastReconcileAttachmentError error = nil
+	for i := range r.currentProgramState.TC.AttachPoints {
+		r.currentAttachPoint = &r.currentProgramState.TC.AttachPoints[i]
+		remove, err := r.reconcileBpfAttachment(ctx, r)
+		if err != nil {
+			r.Logger.Error(err, "failed to reconcile bpf attachment", "index", i)
+			// All errors are logged, but the last error is saved to return and
+			// we continue to process the rest of the attach points so errors
+			// don't block valid attach points.
+			lastReconcileAttachmentError = err
+		}
+
+		if remove {
+			r.Logger.Info("Marking attach point for removal", "index", i)
+			attachPointsToRemove[i] = true
+		}
+	}
+
+	if len(attachPointsToRemove) > 0 {
+		r.Logger.Info("Removing attach points", "attachPointsToRemove", attachPointsToRemove)
+		r.currentProgramState.TC.AttachPoints = r.removeAttachPoints(r.currentProgramState.TC.AttachPoints, attachPointsToRemove)
+	}
+
+	r.updateProgramAttachStatus()
+
+	return lastReconcileAttachmentError
+}
+
+func (r *TcProgramReconciler) updateProgramAttachStatus() {
+	for _, attachPoint := range r.currentProgramState.TC.AttachPoints {
+		if !isAttachSuccess(attachPoint.ShouldAttach, attachPoint.AttachPointStatus) {
+			r.setProgramAttachStatus(bpfmaniov1alpha1.ProgAttachError)
+			return
+		}
+	}
+	r.setProgramAttachStatus(bpfmaniov1alpha1.ProgAttachSuccess)
+}
+
+// removeAttachPoints removes attach points from a slice of attach points based on the keys in the map.
+func (r *TcProgramReconciler) removeAttachPoints(attachPoints []bpfmaniov1alpha1.TcAttachInfoState, attachPointsToRemove map[int]bool) []bpfmaniov1alpha1.TcAttachInfoState {
+	var remainingAttachPoints []bpfmaniov1alpha1.TcAttachInfoState
+	for i, a := range attachPoints {
+		if _, ok := attachPointsToRemove[i]; !ok {
+			remainingAttachPoints = append(remainingAttachPoints, a)
+		}
+	}
+	return remainingAttachPoints
+}
+
+// getExpectedAttachPoints expands *AttachInfo into a list of specific attach
+// points.
+func (r *TcProgramReconciler) getExpectedAttachPoints(ctx context.Context, attachInfo bpfmaniov1alpha1.TcAttachInfo,
+) ([]bpfmaniov1alpha1.TcAttachInfoState, error) {
+	interfaces, err := getInterfaces(&attachInfo.InterfaceSelector, r.ourNode)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get interfaces for TcProgram: %v", err)
+	}
+
+	nodeAttachPoints := []bpfmaniov1alpha1.TcAttachInfoState{}
+
+	if attachInfo.Containers != nil {
+		// There is a container selector, so see if there are any matching
+		// containers on this node.
+		containerInfo, err := r.Containers.GetContainers(
+			ctx,
+			attachInfo.Containers.Namespace,
+			attachInfo.Containers.Pods,
+			attachInfo.Containers.ContainerNames,
+			r.Logger,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get container pids: %v", err)
+		}
+
+		if containerInfo != nil && len(*containerInfo) != 0 {
```

Comment:
nit no need to check the len range will handle that

Reply:
Done.